### PR TITLE
`Development`: Increase code quality for notification target

### PR DESCRIPTION
**Pull request by Malyuk-A**
#### General
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
### Motivation and Context
@krusche asked me to refactor the way notifications had been used to extract their target property to make a valid URL on the server-side for the use in emails.
I refactored the code by replacing explicit NotificationTarget.java classes and a GSON (from JSON->String parser) with a transient database property.

### Description
Previouse Workflow of Notifications -> Email (Similar for Group or SingleUser Notifications)
- NotificationService.java triggered by external REST Resource event (e.g. Exercise updated)
- Notification is created in a Notification Factory
-- a notification has a String "target" property that is intended to store useful information (like elements for a valid URL) in the DB and send them to the client
-- at the client the target property is used to extract a valid URL/Link
-- the target contains also other elements, e.g. messages or exam problem statements 
-- e.g. "{\"problemStatement\":\"PROBLEM STATEMENT\",\"exercise\":3,\"exam\":1,\"entity\":\"exams\",\"course\":1,\"mainPage\":\"courses\"}"
- the target property is created by filling a JsonObject.java with the needed properties and later on converting it to a String.
- There was no issue so far but with the new notification emails the needed arose to also extract valid URLs/Links from the notifications on the serverside (because the emails are made on the server) 
- This has been done by using GSON and explicit java classes (e.g. NotificationTarget.java)

This implementation works but it can be improved.
What did I change:
- I removed the explicit notification target java classes
- I moved the e.g. URL extraction logic into the NotificationTargetService.java
- I added a transient "targetTransient" property to the entity class Notification.java (It is only used during runtime operations and has no connection to the DB)
-- The benefit here is that we no longer need explicit JSON (GSON) parses but we can simply reuse the JsonObject that is created in the NotificationFactories for extracting the needed attributes for the email links.

I also renamed the NotificationTargetService into NotificationTargetProvider (based on one of @krusche's change requests)

### Steps for Testing
Look at the changed code.
Play around with the notification & email system. The goal is to check if the Links / URLs are still valid and no server errors occur.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
NotificationTargetProvider ~ 100%
GroupNotificationService ~ 100%
SingleUserNotificationService ~ 100%
GroupNotificationFactory ~ 90%
SingleUserNotificationFactory > 90%

### CODE REVIEW

**Review by filip-gregurevic — DISMISSED** [review]
Changes look good to me 🚀

**Review by willergeorg — APPROVED** [review]
I tested the following notifications on TS1:

- New course wide post
- Reply on course wide post
- A new announcement
- New text exercise released (Notification and Email)
- Quiz open for practice (Notification and Email)
- New post on quiz exercise
- Reply on a quiz related post

Everything worked fine for me. Approve.

File: src/main/java/de/tum/in/www1/artemis/domain/notification/Notification.java [hunk 1]
```diff
@@ -46,6 +47,14 @@
     @Column(name = "target")
     private String target;
 
+    /**
+     * The String target is created based on a JsonObject
+     * which hold the needed information to build a valid URL/Link
+     * it is used to create Emails without the need to parse it (e.g. via GSON)
+     */
+    @Transient
+    private transient JsonObject targetTransient;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/notification/Notification.java:56]: we want to avoid the use of JsonObject

**Review by filip-gregurevic — APPROVED** [review]
Re-approve 👍🏼

**Review by willergeorg — APPROVED** [review]
Tested every notification except course archiving. Every notification was visible and routing worked for me whenever intended. Re-approve.

**Review by cremertim — APPROVED** [review]
tested on ts1. did not expect any issues while clicking through the notfication system. great work 👍

**Review by DanielKainz — APPROVED** [review]
I tested it on TS1. For me everything worked as expected and I did not encounter any issues. Great work :-)

File: src/main/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProvider.java [hunk 2]
```diff
@@ -1,53 +1,45 @@
...
 
 import de.tum.in.www1.artemis.domain.Course;
 import de.tum.in.www1.artemis.domain.Exercise;
 import de.tum.in.www1.artemis.domain.Lecture;
 import de.tum.in.www1.artemis.domain.ProgrammingExercise;
 import de.tum.in.www1.artemis.domain.metis.Post;
+import de.tum.in.www1.artemis.domain.notification.Notification;
+import de.tum.in.www1.artemis.domain.notification.NotificationTarget;
 
+@Service
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProvider.java:13]: Please do not mix Services with non services. Either this class is a service and is only injected using constructor injection into other services. Then you are not allowed to use the constructor, i.e. no use of

`new NotificationTargetProvider()`

The alternative is, that NotificationTargetProvider is **not** a service, then you use the normal constructor and `new NotificationTargetProvider()` is ok

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProvider.java:13]: Looking at the methods below, I think NotificationTargetProvider should **not** be a service. It is effectively a factory, so we should call it `NotificationTargetFactory`. All methods should be renamed from `getXTarget` into `createXTarget`

File: src/test/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import de.tum.in.www1.artemis.domain.metis.Post;
+import de.tum.in.www1.artemis.domain.notification.Notification;
+import de.tum.in.www1.artemis.domain.notification.NotificationTarget;
+
+public class NotificationTargetProviderTest {
+
+    @Autowired
+    private static NotificationTargetProvider notificationTargetProvider;
+
+    @Mock
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProviderTest.java:27]: I don't understand why you need mocks for entity objects. I think this should not be done!
Please instead configure the objects normally, i.e. set the id in a new object for the specific test and then use the id.

File: src/main/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProvider.java (hunk 2 above)
Comment (Malyuk-A) [reply src/main/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProvider.java:13]: I renamed the provider into a factory and adjusted the method names

File: src/test/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProviderTest.java (hunk 3 above)
Comment (Malyuk-A) [reply src/test/java/de/tum/in/www1/artemis/service/notifications/NotificationTargetProviderTest.java]: Thank you for pointing this out. I replaced the mocked entity types with real instances and added missing assertThat().as("explanation")... 
I did so with multiple test classes -> that is the reason for the big jump in LoC.

**Comment by Malyuk-A** [comment]
I applied the requested changes.
The Codacy issue just states that the naming of the new factory class is wrong, but I followed the legacy naming for the GroupNotificationFactory as well as the SingleUserNotificationFactory, so I think it is fine this way.

**Review by willergeorg — APPROVED** [review]
I retested all notifications on TS1. Still works smoothly. Code changes since my last review seem to be reasonable to me. Re-approve.

**Review by rs-maker — APPROVED** [review]
Tested and it worked correctly. Re approve.

**Review by filip-gregurevic — APPROVED** [review]
Re-approve, code changes look good to me 👍🏼

**Pull request merged**